volatility/framework/plugins/windows/lsadump.py

```diff
@@ -0,0 +1,202 @@
+# This file is Copyright 2020 Volatility Foundation and licensed under the Volatility Software License 1.0
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+import logging
+from volatility.framework import interfaces, renderers
+from volatility.framework.configuration import requirements
+from volatility.framework.renderers import format_hints
+from volatility.framework.layers import intel
+from volatility.plugins.windows.registry import hivelist
+from volatility.plugins.windows import hashdump, poolscanner
+from Crypto.Hash import MD5, SHA256
+from Crypto.Cipher import ARC4, DES, AES
+from struct import unpack, pack
+import collections
+
+vollog = logging.getLogger(__name__)
+
+class Lsadump(interfaces.plugins.PluginInterface):
+    """Dumps lsa secrets from memory"""
+    @classmethod
+    def get_requirements(cls):
+        return [requirements.TranslationLayerRequirement(name = 'primary',
+                                                     description = 'Memory layer for the kernel',
+                                                     architectures = ["Intel32", "Intel64"]),
+            requirements.SymbolTableRequirement(name = "nt_symbols",
+                                                description = "Windows kernel symbols"),
+            requirements.PluginRequirement(name = 'hivelist', plugin = hivelist.HiveList, version = (1, 0, 0))#,
+            ]
+    
+    @classmethod
+    def decrypt_aes(cls, secret, key):
+        """
+        Based on code from http://lab.mediaservice.net/code/cachedump.rb
+        """
+        sha = SHA256.new()
+        sha.update(key)
+        for _i in range(1, 1000 + 1):
+            sha.update(secret[28:60])
+        aeskey = sha.digest()
+
+        data = b""
+        for i in range(60, len(secret), 16):
+            aes = AES.new(aeskey, AES.MODE_CBC, '\x00' * 16)
+            buf = secret[i : i + 16]
+            if len(buf) < 16:
+                buf += (16 - len(buf)) * "\00"
+            data += aes.decrypt(buf)
+
+        return data
+    
+    @classmethod
+    def get_lsa_key(cls, sechive, bootkey, vista_or_later):
+        if not bootkey:
+            return None
+
+        root = sechive.root_cell_offset
+        if not root:
+            return None
+
+        
+        
+        if vista_or_later:
+            policy_key = 'PolEKList'
+        else:
+            policy_key = 'PolSecretEncryptionKey'
+
+        enc_reg_key = sechive.get_key("Policy\\"+policy_key)                    
+        if not enc_reg_key:
+            return None
+        enc_reg_value = next(enc_reg_key.get_values())
+
+
+        if not enc_reg_value:
+            return None
+        
+        obf_lsa_key = sechive.read(enc_reg_value.Data+4, enc_reg_value.DataLength)
+
+        if not obf_lsa_key:
+            return None
+        if not vista_or_later:
+            md5 = MD5.new()
+            md5.update(bootkey.encode('latin1'))
```

Comment:
The bootkey should presumably be in bytes, and again, this calls into hashdump, which was reworked so that bytes were used uniformly as possible.  Please update this to work with the hashdump in the master branch.